### DISCUSSION TITLE
BLE act like serial channel

### DISCUSSION THREAD

**Discussion by mpiffari** [shinyorg/shiny#1179] [Q&A] [answered]
Hi @everybody,
in my spare time I'm developing a _Xamarin mobile application_ project whose job is to exchange data with a BLE device.

In particular, the exchange of information takes place in "_UART style_": it is a request/response approach that relies on the BLE protocol.
In particular, the GATT database of the device is configured in such a way as to receive data on a BLE characteristic "X" and to notify the response on the same BLE characteristic "X".

I wanted to ask you, since I'm starting to get familiar with the RX/Observable pattern now, which is the best approach to write (via the _Write_ function) and then subscribe to _WhenNotificationReceived_ to return the response from the device.

**Comment by aritchie** [comment, marked as answer]
No, always hook the receiver first.  Once that is complete - start your write operations

**Comment by mpiffari** [comment]
Hi @aritchie,
Thanks for your reply.

So, making reference to the APIs of old shiny nuget version ([https://github.com/aritchie/bluetoothle](https://github.com/aritchie/bluetoothle)) an approach like this, with a synchro blocking (forced by high-level interfaces that I need to be compliant with), could make sense?

```
Task.Run(async () =>
{
    timeoutCancellationSource.CancelAfter(TimeSpan.FromMilliseconds(timeout));
    notificationCharact.WhenNotificationReceived().FirstOrDefaultAsync.Subscribe(x =>
    {
        promise.SetResult(x);
    });
    await rx.Write(bytes).ToTask(cancellationToken);
}, cancellationToken).ConfigureAwait(false);
    
response = promise.Task.GetAwaiter().GetResult();
```

**Reply by aritchie** [reply]
Sync blocking is not something you can do successfully with BLE.  If you are doing UART style commands, you won't be guaranteed to get a full response on one transfer packet either.  Usually, the RX is left open for UART.

I'm not supporting/talking anything to do with the old plugin.  This is for Shiny only and the latest stable versions only.  Please be respectful of that.

**Reply by aritchie** [reply]
```csharp
public static IObservable<byte[]?> RequestResponseNotify(this IPeripheral peripheral, byte[] data, string serviceUuid, string charUuid, string? rxCharUuid = null, bool writeWithResponse = true) => Observable.FromAsync<byte[]?>(async ct =>
    {
        var rxTask = peripheral
            .Notify(serviceUuid, rxCharUuid ?? charUuid)
            .Take(1)
            .ToTask(ct);

        await peripheral
            .WriteCharacteristicAsync(serviceUuid, charUuid, data, writeWithResponse, ct)
            .ConfigureAwait(false);

        var result = await rxTask.ConfigureAwait(false);

        return result?.Data;
    });
```
